crates/worker/src/partition_processor_manager/snapshot_task.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+use std::time::SystemTime;
+
+use tokio::sync::{oneshot, watch};
+use tracing::{debug, warn};
+
+use restate_core::worker_api::{SnapshotError, SnapshotResult};
+use restate_partition_store::snapshots::{
+    LocalPartitionSnapshot, PartitionSnapshotMetadata, SnapshotFormatVersion,
+};
+use restate_partition_store::PartitionStoreManager;
+use restate_types::identifiers::{PartitionId, SnapshotId};
+use restate_types::logs::Lsn;
+
+/// Handle to an outstanding [`SnapshotPartitionTask`] that has been spawned, including a reference
+/// to notify the requester.
+pub struct PendingSnapshotTask {
+    pub sender: oneshot::Sender<SnapshotResult>,
+}
+
+/// Creates a partition store snapshot along with Restate snapshot metadata.
+pub struct SnapshotPartitionTask {
+    pub cluster_name: String,
+    pub node_name: String,
+    pub partition_id: PartitionId,
+    pub snapshot_base_path: PathBuf,
+    pub partition_store_manager: PartitionStoreManager,
+    pub archived_lsn_sender: watch::Sender<Option<Lsn>>,
+}
+
+impl SnapshotPartitionTask {
+    pub async fn create_snapshot(self) -> Result<PartitionSnapshotMetadata, SnapshotError> {
+        debug!(
+            partition_id = %self.partition_id,
+            "Creating partition snapshot"
+        );
+
+        let result = create_snapshot_inner(
+            self.partition_store_manager,
+            self.cluster_name,
+            self.node_name,
+            self.partition_id,
+            self.snapshot_base_path,
+            self.archived_lsn_sender,
+        )
+        .await;
+
+        match result {
+            Ok(metadata) => {
+                debug!(
+                    partition_id = %self.partition_id,
+                    snapshot_id = %metadata.snapshot_id,
+                    archived_lsn = %metadata.min_applied_lsn,
+                    "Partition snapshot created"
+                );
+                Ok(metadata)
+            }
+            Err(err) => {
+                warn!(
+                    partition_id = %self.partition_id,
+                    "Failed to create partition snapshot: {}",
+                    err
+                );
```

Comment:
Is this something the users should be concerned about? What are the conditions in which the partition snapshotting fails?

Reply:
I guess metrics could be nice in the future. I was mainly thinking whether a failed snapshot is something that can happen every now and then and where a `warn!` logging would be too alarming because it does not indicate a problematic situation. I guess that ideally we would like to track how often failures occur and if a few happened than raise the awareness of the operator/user. But this is probably for the future.

Reply:
From my understanding about how RocksDB's export column family works, this seems relatively unlikely to happen on a regular basis. Since the new SSTs are generally just hard-links to existing ones, even operating with relatively full disk is unlikely to result in an error here. But I think we'll only know for certain once we have some operational experience 😅